Move logic to infer revision.description to BE

qnkhuat opened the pull request:
### What
move the logic of inferring `revision.description` to BE.

### Motivation
Currently, both FE and BE have code to generate `revision.description`, but apparently FE doesn't use the content generated by BE but infers on its own. 

This duplication is unnecessary and makes adding new revisions a 2-people work. 

So in this PR, we moved the generation of `revision.description` to BE entirely.

This is a preparation to implement revision for dashboard tab ([Epic](https://github.com/metabase/metabase/issues/29502))

### Results screenshots
#### With dashboard
1. Created a dashboard
2. Rename
3. Change the description
3. Add 2 cards
4. Remove 1 card and add a series
5. Rearrange cards
6. Revert

Before
![Screenshot 2023-05-11 at 09 08 49](https://github.com/metabase/metabase/assets/25661381/1c27844e-8d8b-4457-b9f9-5c5cd62811aa)

After
![Screenshot 2023-05-11 at 09 07 38](https://github.com/metabase/metabase/assets/25661381/40ff7c81-95b3-4380-b3ac-5e92dcad8ab1)

#### With questions
1. Create a MBQL question
2. Edit the query
3. change the name
4. change the dipslay
5. change the viz-settings
6. turn to model
7. edit metadata

Before
![Screenshot 2023-05-08 at 21 59 27](https://user-images.githubusercontent.com/25661381/236858768-6cb2c7d8-7d17-4f8b-80ea-a98409182f97.png)

After
![Screenshot 2023-05-08 at 22 00 06](https://user-images.githubusercontent.com/25661381/236858895-a7e1420c-7f91-4844-a616-c217491777b6.png)

Note after merging: create an issue regarding internalization for right-to-left language and language without full-stop.

### Note
A bonus from this change is that we will track revision changes when moving the dashboard to a different collection. 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/30502)
<!-- Reviewable:end -->

qnkhuat committed:
Dashboard Tab entity (#29802)

* add dashboard tab entity

qnkhuat committed:
Dashboard tabs CRUD (#29861)

qnkhuat committed:
merge with master

EmmadUsmani committed:
Merge branch 'master' into feature/dashboard-tab-basic

qnkhuat committed:
migrating revision logic to BE

qnkhuat committed:
handle description for multiple cards add/remove

EmmadUsmani committed:
Merge branch 'master' into feature/dashboard-tab-basic

EmmadUsmani committed:
Merge branch 'master' into feature/dashboard-tab-basic

qnkhuat committed:
test fixes

qnkhuat committed:
adds tests

qnkhuat committed:
make sures events are started so tests work

EmmadUsmani committed:
Dashboard tabs frontend (#30189)

* rename `DashboardHeader.tsx` to `DashboardHeaderView` to avoid collision with `DashboardHeader.jsx`

commit-id:7cdfd86a

* add tabs to dash

commit-id:d3c2fa74

* improve code structure and other fixes

* rename `dashboardtab_id` to `dashboard_tab_id`

* remove explicit `position` logic

* improve code quality

* add renaming functionality

* add two new tabs when creating first tab

* hide tabs if there is only one

* add unit test for `DashboardTabs` component

* refactor

* fix type errors

* add `tabId` to text cards

* add `tabId` to link cards

* add `tabId` to action button

* use disabled state when only one tab remains

* refetch cards with filter value when changing tabs

* add e2e test

* fix broken e2e tests

* add horizontal scrolling

* fix unit test

* fix create tab button being on right

* fix performance issue when switching tabs

* add `ttag` to `tabs.ts`

* use `aria-label` in e2e test helper

* fix type error after rebasing

* making ordered_tabs optional for tests purposes

* fix failing unit tests

* fix failing e2e tests

* fix flaky revision history e2e test

* fix type error after merging feature branch with master and rebasing

* fix revisions unit test

* fix `ActionParmatersInputsForm.tsx` to stop `actions-on-dashboards.cy.spec.js` from flaking

---------

Co-authored-by: Ngoc Khuat <qn.khuat@gmail.com>

EmmadUsmani committed:
remove additional `:maybe` that was added while rebasing

qnkhuat committed:
Merge branch 'master' into feature/dashboard-tab-basic

qnkhuat committed:
add title and has_multile_changes to revision

qnkhuat committed:
fixes potential test flake

qnkhuat committed:
more tests

qnkhuat committed:
remove debug code

qnkhuat committed:
Merge branch 'master' into feature/dashboard-tab-basic

qnkhuat committed:
Merge branch 'feature/dashboard-tab-basic' into ngoc-revision-i18n

qnkhuat committed:
merge master and fix conflicts

qnkhuat committed:
handles collection change description in card and dashboard

qnkhuat commented on src/metabase/models/dashboard.clj:
```diff
@@ -180,7 +180,7 @@
   [_model _id dashboard]
   (-> dashboard
-      (select-keys [:description :name :cache_ttl :auto_apply_filters])
+      (select-keys [:collection_id :description :name :cache_ttl :auto_apply_filters])
```
we probably forgot to do this. We do track collection changes for Card, so we should track for Dashboard too.

qnkhuat committed:
fixes fail to build

EmmadUsmani committed:
use backend for revision history (#30772)

* use BE response for revision history

* update `service.unit.spec.js`

* sort revision and moderation events for question timeline

qnkhuat committed:
test fixes

qnkhuat commented:
```diff
- 1,846
``` 
🥹

qnkhuat commented on src/metabase/models/revision.clj:
```diff
@@ -45,15 +45,17 @@
 
-(defmulti diff-str
-  "Return a string describing the difference between `object-1` and `object-2`."
+(defmulti diff-strs
```
we want this to be a seq of strings now so we can infer `has-multiple-changes` in https://github.com/metabase/metabase/blob/4f15c36457384187ae7fc8271be6fa99e50956ba/src/metabase/models/revision.clj#L111

qnkhuat commented on src/metabase/models/dashboard.clj:
```diff
@@ -214,51 +214,54 @@
-               (< num-cards1 num-cards2) "added a card"
-               (> num-cards1 num-cards2) "removed a card"
-               :else                     "rearranged the cards")))
+               (and
```
now a single revision can create/remove multiple cards at once, so we need to enhance this a bit.

qnkhuat commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,76 @@
+   [toucan2.core :as t2]))
 
-(defn- diff-string* [k v1 v2]
+(defn- diff-strings* [k v1 v2 identifier]
```
this logic try to follow https://github.com/metabase/metabase/blob/d4621d141e441f8fdd8bf16de698e5197016a1ef/frontend/src/metabase/lib/revisions/revisions.js#L100

calherries replied:
Still waiting for this to be resolved, in case you missed

qnkhuat replied:
I already renamed it with your other suggestion, here is the commit https://github.com/metabase/metabase/pull/30502/commits/4f1a002a091e945a797ca327c0fd78fa64e9ef83

calherries replied:
Of course, my bad.

qnkhuat commented on test/metabase/models/revision/diff_test.clj:
```diff
@@ -1,37 +0,0 @@
-(ns metabase.models.revision.diff-test
```
these tests are covered in test/metabase/models/revision_test.clj , this is just a duplication

qnkhuat committed:
create revision directly instead of via API to avoid flake

calherries commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,76 @@
-                                 (diff-string* k (k before) (k after))))
-              build-sentence
-              (str/replace-first #" it " (format " this %s " t))))))
+      (= (count parts) 2) (str (first parts) " " (deferred-tru "and")  " " (second parts) \.)
```
`build-sentence` won't work correctly for right to left languages like arabic. The only solution I can think of is to have each part end in a full stop (in english), and remove the "and" in favour of concatenating the sentences together separated by a line break.

qnkhuat replied:
that's a good observation. maybe we could have a generic solution for this.

Do you think this is a must for this PR? it's an existing behavior on FE btw.

calherries replied:
True, if it's existing on the FE I guess it could make sense to solve it in another PR.

calherries commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,76 @@
   [parts]
   (when (seq parts)
     (cond
       (= (count parts) 1) (str (first parts) \.)
```
Not all languages use a full-stop to end a sentence, e.g. [Japanese](https://en.wikipedia.org/wiki/Japanese_punctuation). The full stop should included in each element of `parts`

qnkhuat replied:
that's a good catch. but I think we can remove the `.`. we currently don't have in on FE anyway.

calherries commented on src/metabase/models/revision.clj:
```diff
@@ -45,15 +45,17 @@
   [model o1 o2]
   (when-let [[before after] (data/diff o1 o2)]
-    (diff-string (name model) before after)))
+    (diff-strings (name model) before after)))
```
This was an issue before, but the naming of `diff-strs` and `diff-strings` is confusing. I think we should rename `diff-strs` to `diff-strings`, `diff-strings` to `diff-strings*`, and `diff-strings*` to `diff-string`. Then `diff-strings*` is clearly a helper function for `diff-strings`.

qnkhuat replied:
that makes sense, updated here 4f1a002a09

calherries commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,76 @@
+  [model before after]
+  (let [ks (keys (or after before))]
+    (filter identity
+            (map-indexed (fn [i k]
+                           (diff-strings* k (k before) (k after)
+                                          (if (zero? i) (deferred-tru "this {0}" model) (deferred-tru "it")))) ks))))
```
`ks` should be on a new line
```suggestion
            (map-indexed (fn [i k]
                           (diff-strings* k (k before) (k after)
                                          (if (zero? i) (deferred-tru "this {0}" model) (deferred-tru "it"))))
                         ks))))
```

calherries replied:
This is probably already a problem, but "this {0}" and "it" won't internationalize well. For example, "this dashboard" should translate to "la tableau de bord" and "this card" translates to "cette carte". Not sure if this is a problem worth solving right now though 🤷

qnkhuat replied:
the `model` won't be translated. you can see this test for example https://github.com/metabase/metabase/blob/88fadceb53bb7bf7301b0395fc2affdc2961bf2b/test/metabase/api/revision_test.clj#L380

calherries replied:
Ha, I didn't notice that before. Shouldn't it be translated though?

qnkhuat replied:
good question. I think it should 6430fac541

calherries commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,76 @@
+
+(defn diff-strings
+  "Create a seq of string describing how `o1` is different from `o2`.
+  The directionality of the statement should indicate that `o1` changed into `o2`."
```
This docstring is confusing because `o1` and `o2` refer to the arguments of `diff-strs`, not this function.

A suggestion: any reader of `diff-strings` needs to understand that `before` and `after` are results from `data/diff`. This could be avoided if (1) `diff-strings` was inlined in `diff-strs` `:default` body, or (2) we refactored so `data/diff` was called inside `diff-strings` instead. Then `diff-strs` `:default` body would simply be:
```
(defmethod diff-strs :default
  [model o1 o2]
  (when (not= o1 o2)
    (diff-strings (name model) o1 o2)))
```
But then at that point you might as well inline `diff-strings`, because this function is not doing much on its own.

qnkhuat replied:
I moved the diff into `diff-strings*` 4f6f0bd02e

calherries commented on src/metabase/models/dashboard.clj:
```diff
@@ -214,51 +214,54 @@
+                                    (deferred-tru "modified the series on card {0}" (get-in prev-dashboard [:cards idx :card_id]))))))]
+    (-> [(when-let [default-description (build-sentence ((get-method revision/diff-strs :default) Dashboard prev-dashboard dashboard))]
+           (cond-> default-description
+             (str/ends-with? default-description ".") (subs 0 (dec (count default-description)))))
```
`build-sentence` should always return a sentence ending in a full-stop, so this `cond->` is unnecessary. In another of my [comments](https://github.com/metabase/metabase/pull/30502/files#r1195922590) I say the full-stop won't be internationalized well, so this will have to change anyway. You'll need to have another way of removing the end of the sentence marker, but this is tricky. The only workaround I can imagine is not trying to construct sentences in every language and instead have each change as separate sentences.

qnkhuat replied:
as discussed this should be done in a separated PR, let's focus this PR on moving where the logic is handled.
I'll create an issue once this is merged.

calherries reviewed (changes requested):
The main problem is that we're now translating the description generated on the BE into many languages, but still trying to construct sentences with english-like grammar rules. I think we should drop attempting to construct sentences, because it's too hard to generalise this across all languages.

(Edit: since this is an existing problem we should fix this in another PR. I still have other suggestions)

calherries commented on test/metabase/models/revision_test.clj:
```diff
@@ -280,3 +286,56 @@
+                                                      :is_creation  false}))))
+
+     (testing "multiple changes"
+       (is (= {:title                (deferred-tru "edited this.")
```
```suggestion
       (is (= {:title                "edited this."
```

qnkhuat committed:
reverted to an earlier revision => reverted to an earlier version

qnkhuat committed:
translate the model name too

qnkhuat committed:
Rename functions:
- diff-strs -> diff-strings
- diff-strings -> diff-strings*
- diff-strings* -> diff-string

qnkhuat commented on src/metabase/models/revision.clj:
```diff
@@ -83,12 +85,37 @@
+  [model prev-revision revision]
+  (cond
+    (:is_creation revision)  [(deferred-tru "created this")]
+    (:is_reversion revision) [(deferred-tru "reverted to an earlier revision")]
```
good catch, I changed to version [4d343c7](https://github.com/metabase/metabase/pull/30502/commits/4d343c7d28acb924bcab6a28de0affed0b53eac2)

qnkhuat committed:
move the diff to diff-strings*

qnkhuat commented on frontend/src/metabase/components/Timeline/Timeline.tsx:
```diff
@@ -0,0 +1,107 @@
+              <Timestamp dateTime={timestamp}>
+                {getRelativeTime(timestamp)}
+              </Timestamp>
+              {revision?.has_multiple_changes && <div>{description}</div>}
```
I don't think BE should return `nil` description since before this PR it's the only field, `title` is currently added.

And since this is a generic API, we might break others thing if we do that.

calherries replied:
Hmm. I'm trying to figure out how to avoid having a duplicate `title` and `description`, and logic distributed between the FE and BE determining what gets displayed where.

What if we kept `has_multiple_changes` but removed `title` from the API response body, and displayed `edited this.` as the title if `has_multiple_changes` is true? Otherwise the title would be the description.

qnkhuat replied:
that would work! should have thought of this first, I think FE could happily handle that.

qnkhuat replied:
I made the change here 5aafd02835

@EmmadUsmani I made some changes to FE to test things out, feel free to edit it if it doesn't look right for you :)

calherries replied:
The change looks good to me

calherries commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,86 @@
+  [model o1 o2]
+  (when-let [[before after] (data/diff o1 o2)]
+    (let [ks         (keys (or after before))
+          model-name (or (model-str->i18n-str model) model)]
```
Shouldn't we require the model to be translated?
```suggestion
          model-name (model-str->i18n-str model)]
```

`model-str->i18n-str`'s body should be a case statement and return an error if the model is not found.

qnkhuat replied:
yea case might be tighter, how abt this? 5cb31b1817

qnkhuat committed:
use case instead of map for model-str->i18n-str

calherries commented on src/metabase/models/revision/diff.clj:
```diff
@@ -1,50 +1,92 @@
+    "Metric"    (deferred-tru "Metric")
+
+    ;; for testing
+    "NonExistModel" (deferred-tru "NonExistModel")
```
I think you should use `with-redefs` for this in testing, and not complicate production code

qnkhuat replied:
5aafd02835 yep

qnkhuat committed:
remove title from API response

calherries reviewed (approved):
BE looks good now

calherries commented on test/metabase/models/revision_test.clj:
```diff
@@ -13,6 +15,17 @@
 
 (models/defmodel ^:private FakedCard :report_card)
 
+(use-fixtures :each (fn [thunk]
+                      (with-redefs [metabase.models.revision.diff/model-str->i18n-str (fn [model-str]
+                                                                                        (case model-str
+                                                                                          "Dashboard" (deferred-tru "Dashboard")
+                                                                                          "Card"      (deferred-tru "Card")
+                                                                                          "Segment"   (deferred-tru "Segment")
+                                                                                          "Metric"    (deferred-tru "Metric")
+                                                                                          "NonExistModel" (deferred-tru "NonExistModel")
+                                                                                          "FakeCard"      (deferred-tru "FakeCard")))]
+                        (thunk))))
```
Does this mean "NonExistModel" and "FakeCard" will be included in the strings that need translation? Or are they excluded because this is in a test namespace?

qnkhuat replied:
ah we just need it for the model dispatch, it can be included if the object is used .

calherries replied:
I don't understand. Is that a yes, or no?

For clarity I'm asking if it should be this instead, to avoid translators trying to translate "NonExistModel":
```clojure
(use-fixtures :each (fn [thunk]
                      (with-redefs [metabase.models.revision.diff/model-str->i18n-str (fn [model-str]
                                                                                        (case model-str
                                                                                          "Dashboard" (deferred-tru "Dashboard")
                                                                                          "Card"      (deferred-tru "Card")
                                                                                          "Segment"   (deferred-tru "Segment")
                                                                                          "Metric"    (deferred-tru "Metric")
                                                                                          "NonExistModel" "NonExistModel"
                                                                                          "FakeCard"      "FakeCard"))]
                        (thunk))))
```

qnkhuat replied:
currently not, and yes we should not translate it. I'll update the code

qnkhuat committed:
do not translate test model

EmmadUsmani committed:
rename var `title_text` to `titleText`

alxnddr commented:
I noticed that it does not behave correctly in some cases:
<video src="https://github.com/metabase/metabase/assets/14301985/17651c9f-18a7-41b1-ae31-d31445b5e01a" />
Can you please double-check and cover this and other potentially incorrect cases with specs?

qnkhuat commented:
@alxnddr thanks for testing! 
but this behavior exists on master as well. This PR is purely about moving the logic to BE so I think we should create a separated issue and fix it later.

qnkhuat committed:
Merge branch 'master' into ngoc-revision-i18n

alxnddr commented:
@qnkhuat I posted it because I did not see it on `master` so I thought it is a regression in this PR:
<video src="https://github.com/metabase/metabase/assets/14301985/09652dd9-a164-44d7-a1fc-e16bff12d69e" />

qnkhuat commented:
That's weird... would you mind checking again?

I'm pretty sure it happens on master too, in fact, I have a fix for it here https://github.com/metabase/metabase/pull/30868

Note that the event system happens in async, so it might takes some delay to see the complete revision, a page refresh will probably do it.

https://github.com/metabase/metabase/assets/25661381/4182942f-3509-4868-81a5-502ea9252433

ranquild commented on frontend/src/metabase/plugins/index.ts:
```diff
@@ -19,13 +19,11 @@ import type {
   User,
 } from "metabase-types/api";
 import type { AdminPathKey, State } from "metabase-types/store";
+import type { TimelineEvent } from "metabase/components/Timeline";
```
Could we move this type here, in `plugins/index.ts` instead of using one from the component? Components should depend on plugins but not wise-versa. And another thing - we already have an entity called `TimelineEvent`, could we name this one differently? Like `RevisionEvent` or whatever

EmmadUsmani replied:
Yup I can do those both

ranquild commented on frontend/src/metabase/components/Timeline/Timeline.tsx:
```diff
@@ -0,0 +1,107 @@
+const HALF_ICON_SIZE = ICON_SIZE / 2;
+
+type Icon = string | { name: string; color: string } | Record<string, never>;
+export type TimelineEvent = {
```
`metabase/components` is a deprecated folder. If we're creating a domain-specific component that should be reused across features, it should go to `metabase/common/components`

ranquild commented on frontend/src/metabase/components/Timeline/Timeline.tsx:
```diff
@@ -0,0 +1,107 @@
+  Timestamp,
+} from "./Timeline.styled";
+
+const ICON_SIZE = 16;
```
We usually use `rem` for spacing and not pixels.

Another thing is that it looks like something styles-specific, so we could move these constants to styled components and use them there, and remove props like `leftShift`

EmmadUsmani replied:
Yeah I think this is leftover from the old component implementation, I will remove the props

npfitz commented on frontend/src/metabase/components/Timeline/Timeline.styled.tsx:
```diff
@@ -19,17 +22,17 @@ export const TimelineItem = styled.li`
-export const ItemIcon = styled(Icon)`
+export const StyledIcon = styled(Icon)`
   position: relative;
   color: ${props => (props.color ? color(props.color) : color("text-light"))};
```
If all we're really doing is changing the color, the `Icon` component accepts color as a prop. We may not need a styled component here

npfitz commented on frontend/src/metabase/dashboard/components/DashboardInfoSidebar.tsx:
```diff
@@ -33,20 +33,28 @@ interface DashboardInfoSidebarProps {
 
-const DashboardInfoSidebar = ({
+export const DashboardInfoSidebar = _.compose(
+  Revision.loadList({
```
Do we want to try building a `use-revision-list` hook rather than wrapping the component?

qnkhuat committed:
Merge branch 'master' into ngoc-revision-i18n

EmmadUsmani commented on frontend/src/metabase/query_builder/components/QuestionActivityTimeline.tsx:
```diff
@@ -1,59 +1,53 @@
+}
 
-export default _.compose(
+export const QuestionActivityTimeline = _.compose(
```
Yup I can make hooks for both

EmmadUsmani committed:
address comments

EmmadUsmani commented:
@ranquild @npfitz Thank you both for the feedback! I've made changes to address your comments if you guys want to have another look

qnkhuat committed:
stringify the display name

calherries commented:
@qnkhuat Looks like there's at least one test failing due to the BE changes: https://github.com/metabase/metabase/actions/runs/5052809604/jobs/9066196959?pr=30502#step:12:436

Before, the description was "You created this"
Now, it is "You created this."

![image](https://github.com/metabase/metabase/assets/39073188/8e565c03-b54b-4574-b1e0-c70caf5308b8)

I think you should just update this test and make use of a full stop to be consistent for all descriptions and titles, so changing "You edited this" to be "You edited this.".

EmmadUsmani committed:
fix revert button not working

EmmadUsmani committed:
add loading wrapper to fix unit test

EmmadUsmani committed:
fix tests

EmmadUsmani committed:
Merge branch 'master' into ngoc-revision-i18n

EmmadUsmani committed:
fix type error

EmmadUsmani committed:
fix QuestionActivityTimeline test

qnkhuat committed:
add a dot for edited this

qnkhuat commented:
@calherries thanks for pointing that out, the title is now inferred on the FE, so the fix should be on FE. I made the change here https://github.com/metabase/metabase/pull/30502/commits/d9ec970cf829c6e5680c3f084827c0753cc72cc7

qnkhuat committed:
Merge branch 'master' into ngoc-revision-i18n

EmmadUsmani committed:
Merge branch 'master' into ngoc-revision-i18n

qnkhuat committed:
fix failing test

qnkhuat committed:
Merge branch 'master' into ngoc-revision-i18n

qnkhuat committed:
fix conflict

qnkhuat committed:
remove dispatched got defined twice

qnkhuat committed:
fix revert dashboard creates new revision (#30868)
